src/internal/Confirm.tsx

```diff
@@ -0,0 +1,165 @@
+import * as React from 'react';
+import { Drop, DropProps } from 'grommet';
+import styled from 'styled-components';
+import { px } from '../utils';
+import { Box, Button } from '..';
+import TooltipArrow, { TooltipArrowProps } from './TooltipArrow';
+
+const StyledDrop = styled(Drop)`
+	border: 0;
+`;
+
+const StyledDropInner = styled.div`
+	margin: 5px;
+	background: white;
+	box-shadow: ${(props) => '1px 1px 5px' + props.theme.colors.gray.light};
+	border-radius: ${(props) => px(props.theme.radius)};
+	border: ${(props) => '1px solid ' + props.theme.colors.gray.main};
+`;
+
+const getDropAlign = (
+	placement: TooltipArrowProps['placement'],
+): DropProps['align'] => {
+	switch (placement) {
+		case 'bottom':
+			return {
+				top: 'bottom',
+			};
+		case 'top':
+			return {
+				bottom: 'top',
+			};
+		case 'left':
+			return {
+				right: 'left',
+			};
+		case 'right':
+			return {
+				left: 'right',
+			};
+	}
+};
+
+export interface ConfirmOptions {
+	text: string;
+	placement?: TooltipArrowProps['placement'];
+}
+
+interface ConfirmProps extends DropProps, ConfirmOptions {
+	targetRef: React.RefObject<HTMLElement>;
+	onOK: () => any;
+	onCancel: () => any;
+}
+
+const Confirm = ({
+	text,
+	targetRef,
+	onOK,
+	onCancel,
+	placement,
+	...rest
+}: ConfirmProps) => {
+	const dropAlign = React.useMemo(() => {
+		return getDropAlign(placement);
+	}, [placement]);
+
+	return (
+		open &&
+		targetRef.current && (
+			<StyledDrop
+				{...rest}
+				target={targetRef.current}
+				onClickOutside={onCancel}
+				onEsc={onCancel}
+				align={dropAlign}
+				plain
+			>
+				<TooltipArrow placement={placement} />
+				<StyledDropInner>
+					<Box mx={2} mt={2} fontSize={3}>
+						{text}
+					</Box>
+					<Box>
+						<Button m={2} size="small" onClick={onCancel}>
+							Cancel
+						</Button>
+						<Button m={2} size="small" primary onClick={onOK}>
+							OK
+						</Button>
+					</Box>
+				</StyledDropInner>
+			</StyledDrop>
+		)
+	);
+};
+
+Confirm.defaultProps = {
+	placement: 'top',
+};
+
+export const useForwardedRef = <T extends {}>(
+	ref: ((instance: T | null) => void) | React.MutableRefObject<T | null> | null,
+) => {
+	const innerRef = React.useRef<T>(null);
+
+	React.useEffect(() => {
+		if (!ref) {
+			return;
+		}
+
+		if (typeof ref === 'function') {
+			ref(innerRef.current);
+		} else {
+			ref.current = innerRef.current;
+		}
+	});
+
+	return innerRef;
+};
+
+export function withConfirm<T extends HTMLElement>(
+	Base: React.ForwardRefExoticComponent<any>,
+) {
+	return React.forwardRef<
+		T,
+		{
+			onClick: (e: React.SyntheticEvent<T>) => any;
+			confirmOptions: ConfirmOptions;
+		}
+	>(({ onClick, confirmOptions, ...rest }, ref) => {
```

Comment:
As a general comment, It can make it more readable if you define the types separately so that not so much of the inline code is taken up with complex type definitions. 

e.g. 
```typescript
type Confirm<T> = {
			onClick: (e: React.SyntheticEvent<T>) => any;
			confirmOptions: ConfirmOptions;
}
...
React.forwardRef<T, Confirm<T>>
```